source/MaterialXGenShader/Util.cpp

```diff
@@ -608,4 +608,147 @@ void findRenderableElements(const DocumentPtr& doc, std::vector<TypedElementPtr>
     }
 }
 
+ValueElementPtr findNodeDefChild(const string& path, DocumentPtr doc, const string& target)
+{
+    if (path.empty() || !doc)
+    {
+        return nullptr;
+    }
+    ElementPtr pathElement = doc->getDescendant(path);
+    if (!pathElement || pathElement == doc)
+    {
+        return nullptr;
+    }
+    ElementPtr parent = pathElement->getParent();
+    if (!parent || parent == doc)
+    {
+        return nullptr;
+    }
+
+    // Note that we must cast to a specific type derived instance as getNodeDef() is not
+    // a virtual method which is overridden in derived classes.
+    NodeDefPtr nodeDef = nullptr;
+    ShaderRefPtr shaderRef = parent->asA<ShaderRef>();
+    if (shaderRef)
+    {
+        nodeDef = shaderRef->getNodeDef();
+    }
+    else
+    {
+        NodePtr node = parent->asA<Node>();
+        if (node)
+        {
+            nodeDef = node->getNodeDef(target);
+        }
+    }
+    if (!nodeDef)
+    {
+        return nullptr;
+    }
+
+    // Use the path element name to look up in the equivalent element
+    // in the nodedef as only the nodedef elements contain the information.
+    const std::string& valueElementName = pathElement->getName();
+    ValueElementPtr valueElement = nodeDef->getChildOfType<ValueElement>(valueElementName);
+
+    return valueElement;
+}
+
+unsigned int getUIProperties(const ValueElementPtr nodeDefElement, UIProperties& uiProperties)
+{
+    if (!nodeDefElement)
+    {
+        return 0;
+    }
+
+    unsigned int propertyCount = 0;
+    uiProperties.uiName = nodeDefElement->getAttribute(ValueElement::UI_NAME_ATTRIBUTE);
+    if (!uiProperties.uiName.empty())
+        propertyCount++;
+
+    uiProperties.uiFolder = nodeDefElement->getAttribute(ValueElement::UI_FOLDER_ATTRIBUTE);
+    if (!uiProperties.uiFolder.empty())
+        propertyCount++;
+
+    if (nodeDefElement->isA<Parameter>())
+    {
+        string enumString = nodeDefElement->getAttribute(ValueElement::ENUM_ATTRIBUTE);
```

Comment:
Since this is done as an outside util. Spend some time to return some readable values vs just a long string seperated by commas.